### DISCUSSION TITLE
Use child in fix?

### DISCUSSION THREAD

**Discussion by hobofan** [ast-grep/ast-grep#480] [Q&A] [answered]
Hey there 👋 

I'm trying to create a rule with a fix, that replaces the matched node with a child of the matched node. However, there doesn't seem to be a way (in the YAML syntax) to extract/select a child/field of a matched node. Is that correct?

I discovered the "transform" mechanism, which seems to take some steps towards that.

## Concrete example

Goal: Lift up the `object.(FooBar|BarFoo).iri` to replace the matched `object` (pseudo-syntax)

Sample input:
```
const output = Thing.doThing(testInput, {
            known: [
                {
                    FooBar: {
                        iri: {
                            _type: 'IRI',
                            string: 'urn:abc',
                        },
                        annotations: [],
                    },
                },
                {
                    FooBar: {
                        iri: {
                            _type: 'IRI',
                            string: 'urn:cde',
                        },
                        annotations: [],
                    },
                },
                {
                    BarFoo: {
                        iri: someFn1(),
                        annotations: [],
                    },
                },
                {
                    FooBar: {
                        iri: someFn2(),
                        annotations: [],
                    },
                },
            ],
        })
```

Sample output:
```
const output = Thing.doThing(testInput, {
            known: [
                {
                    _type: 'IRI',
                    string: 'urn:abc',
                },
                {
                    _type: 'IRI',
                    string: 'urn:cde',
                },
                someFn1(),
                someFn2(),
            ],
        })
```

**Comment by HerringtonDarkholme** [comment, marked as answer]
Hi, thanks for using ast-grep. You can use relational rule to extract child.

```yaml
rule:
  any:
    - pattern: '{FooBar: $CHILD}'
    - pattern: '{BarFoo: $CHILD}'
  has:
    pattern: '{ iri: $B, $$$ }'
    stopBy: end
fix: $B
```

<img width="641" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/59793a28-c740-4545-813f-0252434f0531">

[link](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiQ29uZmlnIiwibGFuZyI6ImphdmFzY3JpcHQiLCJxdWVyeSI6IntGb29iYX0iLCJyZXdyaXRlIjoibG9nZ2VyLmxvZygkTUFUQ0gpIiwiY29uZmlnIjoiIyBZQU1MIFJ1bGUgaXMgbW9yZSBwb3dlcmZ1bCFcbiMgaHR0cHM6Ly9hc3QtZ3JlcC5naXRodWIuaW8vZ3VpZGUvcnVsZS1jb25maWcuaHRtbCNydWxlXG5ydWxlOlxuICBhbnk6XG4gICAgLSBwYXR0ZXJuOiAne0Zvb0JhcjogJENISUxEfSdcbiAgICAtIHBhdHRlcm46ICd7QmFyRm9vOiAkQ0hJTER9J1xuICBoYXM6XG4gICAgcGF0dGVybjogJ3sgaXJpOiAkQiwgJCQkIH0nXG4gICAgc3RvcEJ5OiBlbmRcbmZpeDogJEIiLCJzb3VyY2UiOiJjb25zdCBvdXRwdXQgPSBUaGluZy5kb1RoaW5nKHRlc3RJbnB1dCwge1xuICAgICAgICAgICAga25vd246IFtcbiAgICAgICAgICAgICAgICB7XG4gICAgICAgICAgICAgICAgICAgIEZvb0Jhcjoge1xuICAgICAgICAgICAgICAgICAgICAgICAgaXJpOiB7XG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgX3R5cGU6ICdJUkknLFxuICAgICAgICAgICAgICAgICAgICAgICAgICAgIHN0cmluZzogJ3VybjphYmMnLFxuICAgICAgICAgICAgICAgICAgICAgICAgfSxcbiAgICAgICAgICAgICAgICAgICAgICAgIGFubm90YXRpb25zOiBbXSxcbiAgICAgICAgICAgICAgICAgICAgfSxcbiAgICAgICAgICAgICAgICB9LFxuICAgICAgICAgICAgICAgIHtcbiAgICAgICAgICAgICAgICAgICAgRm9vQmFyOiB7XG4gICAgICAgICAgICAgICAgICAgICAgICBpcmk6IHtcbiAgICAgICAgICAgICAgICAgICAgICAgICAgICBfdHlwZTogJ0lSSScsXG4gICAgICAgICAgICAgICAgICAgICAgICAgICAgc3RyaW5nOiAndXJuOmNkZScsXG4gICAgICAgICAgICAgICAgICAgICAgICB9LFxuICAgICAgICAgICAgICAgICAgICAgICAgYW5ub3RhdGlvbnM6IFtdLFxuICAgICAgICAgICAgICAgICAgICB9LFxuICAgICAgICAgICAgICAgIH0sXG4gICAgICAgICAgICAgICAge1xuICAgICAgICAgICAgICAgICAgICBCYXJGb286IHtcbiAgICAgICAgICAgICAgICAgICAgICAgIGlyaTogc29tZUZuMSgpLFxuICAgICAgICAgICAgICAgICAgICAgICAgYW5ub3RhdGlvbnM6IFtdLFxuICAgICAgICAgICAgICAgICAgICB9LFxuICAgICAgICAgICAgICAgIH0sXG4gICAgICAgICAgICAgICAge1xuICAgICAgICAgICAgICAgICAgICBGb29CYXI6IHtcbiAgICAgICAgICAgICAgICAgICAgICAgIGlyaTogc29tZUZuMigpLFxuICAgICAgICAgICAgICAgICAgICAgICAgYW5ub3RhdGlvbnM6IFtdLFxuICAgICAgICAgICAgICAgICAgICB9LFxuICAgICAgICAgICAgICAgIH0sXG4gICAgICAgICAgICBdLFxuICAgICAgICB9KSJ9)

**Reply by hobofan** [reply]
Ah, great! Thanks for the swift reply!

From most of the examples found in the docs, I was under the impression that only atomic rules would bind metavariables.

**Reply by HerringtonDarkholme** [reply]
Thanks for the suggestion! I will add it as a tip to the doc! https://github.com/ast-grep/ast-grep.github.io/issues/161
